---
title: "Tidy Meta Profiles using Bioconductor and the Tidyverse"
author: "Malte Thodberg"
package: TeMPO
abstract: >
  The TeMPO R/Bioconductor package provides fast and easy to use functions for generating _meta profiles_ (also known as _meta genes_, _average profiles_ or _average footprints_) over one or more sets of genomic locations and signals. TeMPO uses highly efficient, but difficult to use Bioconductor S4-classes to calculcate summaries, but outputs results as ordinary data.frames or tibbles for easy and flexible plotting with the popular ggplot2 package. TeMPO thereby provides the glue for working with Bioconductor-objects in the tidyverse. Additional meta-profile functionalites such as stranded meta-profiles, outlier-trimming and alternative summary functions are available for advanced users.
output: 
    BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{TeMPO}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}  
---

# Installation

Install the most recent stable version from Bioconductor:

```{r bioconductor, eval=FALSE}
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("TeMPO")
```

And load `TeMPO`:

```{r library, results='hide', message=FALSE}
library(TeMPO)
```

Alternatetively, you can install the development version directly from GitHub using `devtools`:

```{r github, eval=FALSE}
devtools::install_github("MalteThodberg/TeMPO")
```

# Getting help and contact

For general questions about the usage of TeMPO, use the [official Bioconductor support forum](https://support.bioconductor.org) and tag your question "TeMPO". We strive to answer questions as quickly as possible.

For technical questions, bug reports and suggestions for new features, we refer to the [TeMPO github page](https://github.com/MalteThodberg/TeMPO/issues)

# Usage examples

## Inputs and outputs

TeMPO can calculate the summarized signal (e.g. the average counts-per-millon) across a set of genomic regions (e.g. Transcription Start Sites (TSSs) and enhancers). TeMPO needs the following input:

- One or more sets of genomic sites as a GRanges or GRangesList.
- One or more sets of genome-wide signals as a BigWigFile or BigWigFilelist (or RleList or RleListList if the genome is very small and/or the genome-wide signal is very sparse).

TeMPO will output the result in "tidy" format as a tibble, making it easy to further analyze or plot the data using ggplot2 and the tidyverse packages.

## Example data sets

TeMPO example below use example data for hela cells (hg19) chromosomes 20-22 from the following sources:

- CAGE data from "Nuclear stability and transcriptional directionality separate functionally distinct RNA species" by Andersson et al. preprocessed using the CAGEfightR package.
- ChIP-Seq and DNase-Seq data from The Roadmap Epigenomics Project (obtained via the AnnotationHub package with IDs: AH32877, AH32879, AH32881, AH32884).

First, we load the needed packages:

```{r, message=F, warning=F}
library(tidyverse)
theme_set(theme_light())
```

`CAGE_clusters` contain the location of TSSs and enhancer found using the CAGEfightR package. TSSs and enhancers are expected to show different epigenetic modifications, reflecting their different function:
```{r}
data("CAGE_clusters")
CAGE_clusters
```

`CAGE_plus` and `CAGE_minus` contains the pooled CAGE signal across wildtype cells (WT) and exosome knock-out cell (KD). The exosome knock-down is expected to stabilize eRNA transcripts produced from enhancers that are normally quickly degraded. As the CAGE data is very sparse, it can be loaded directly into memory as an `RleListList`:

```{r}
data("CAGE_plus")
data("CAGE_minus")
CAGE_plus
```

We can obtain epigenetic data for comparison to the CAGE data using the The Roadmap Epigenomics Project. As ChIP-Seq and DNase-signal is not sparse, we will store these signals as BigWigFile-objects rather than load them into memory:

```{r}
# Locate the bigwig files and name them
bw_files <- system.file("extdata", package="TeMPO") %>% 
    list.files(full.names=TRUE)

names(bw_files) <- bw_files %>%
    basename() %>%
    tools::file_path_sans_ext()
    
# Save them as a BigWigFileList
ChIP_Seq <- bw_files %>%
    lapply(BigWigFile) %>%
    as("List")
```

```{r, eval=FALSE}
# NOTE: The first time you run this, it will take several minutes for AnnotationHub to download the file. After that, AnnotationHub will simply fetch the file from cache, as is done here:__
ah <- AnnotationHub()
ChIP_Seq <- list(DNase="AH32877",
     H3K4me1="AH32879",
     H3K4me3="AH32881",
     H3K27ac="AH32884") %>%
    lapply(function(x) ah[[x]]) %>%
    as("List")
```

```{r, eval=FALSE}
# Get metadata
ah <- AnnotationHub()
meta_data <- ah[c("AH32877", "AH32879", "AH32881", "AH32884")] %>% 
    mcols

# Make BigWigFileList of URLs and set names
ChIP_Seq <- meta_data$sourceurl %>%
    lapply(BigWigFile) %>%
    as("List")

names(ChIP_Seq) <- str_remove_all(meta_data$title, 
                                  pattern = "E117-|.fc.signal.bigwig")
```

Below are examples of generating metaprofiles for different combinations of input.

## Single set of sites & single genome-wide signal

A simple example of (unstranded) DNAse-signal around promoters:
```{r}
promoters_only <- subset(CAGE_clusters, txType == "promoter")

SS1 <- tidyMetaProfile(sites = promoters_only, 
                      forward=ChIP_Seq$DNase, reverse=NULL,
                      upstream=300, downstream=300)

head(SS1)

ggplot(SS1, aes(x=pos0, y=sense)) + 
    geom_line(alpha=0.75) +
    geom_vline(xintercept=0, alpha=0.75, linetype="dotted") +
    labs(x="Basepair position relative to center",
         y="Average DNase signal")
```

Classic example of bidirectional transcription of eRNAs at enhancers shown using a stranded meta-profile:
```{r}
enhancers_only <- subset(CAGE_clusters, clusterType == "enhancer")

SS2 <- tidyMetaProfile(sites = enhancers_only, 
                      forward=CAGE_plus$WT, reverse=CAGE_minus$WT,
                      upstream=250, downstream=250)

head(SS2)

SS2 %>%
    gather(key="direction", value="score", sense, anti, factor_key=TRUE) %>%
    ggplot(aes(x=pos0, y=score, color=direction)) + 
    scale_color_brewer("Direction", palette="Set1") +
    geom_line(alpha=0.75) +
    geom_vline(xintercept=0, alpha=0.75, linetype="dotted") +
    labs(x="Basepair position relative to center",
         y="Average CAGE signal")
```

## Single set of sites & multiple genome-wide signal

Multiple different epigentic signals across a single set of sites:
```{r}
SM <- tidyMetaProfile(sites = promoters_only, 
                      forward=ChIP_Seq, reverse=NULL,
                      upstream=300, downstream=300)

head(SM)

ggplot(SM, aes(x=pos0, y=sense, color=signal)) + 
    geom_line(alpha=0.75) +
    geom_vline(xintercept=0, alpha=0.75, linetype="dotted") +
    labs(x="Basepair position relative to center",
         y="Average CAGE signal")
```

## Multiple sets of sites & single genome-wide signal

H3K27ac at CAGE-defined TSSs at different positions in genes:
```{r}
by_txType <- CAGE_clusters %>%
    subset(clusterType == "TSS" & txType %in% c("promoter", 
                                                "fiveUTR", 
                                                "proximal", 
                                                "intron")) %>%
    splitAsList(.$txType, drop=TRUE)

MS <- tidyMetaProfile(sites = by_txType, 
                      forward=ChIP_Seq$DNase, reverse=NULL,
                      upstream=300, downstream=300)

head(MS)

ggplot(MS, aes(x=pos0, y=sense, color=sites)) + 
    geom_line(alpha=0.75) + 
    geom_vline(xintercept = 0, linetype="dotted", alpha=0.75) +
    scale_color_brewer("Genic Context", palette="Set2") +
    labs(x="Relative position from center", 
         y="Average H3K27ac Signal")
```

## Multiple sets of sites & multiple genome-wide signal

Meta-profile showing the difference in epigentic modifications at TSSs vs enhancers:
```{r}
by_clusterType <- split(CAGE_clusters, CAGE_clusters$clusterType)

MM1 <- tidyMetaProfile(sites = by_clusterType, 
                      forward=ChIP_Seq, reverse=NULL,
                      upstream=300, downstream=300)

head(MM1)

ggplot(MM1, aes(x=pos0, y=sense, color=sites)) + 
    geom_line(alpha=0.75) + 
    facet_grid(signal~., scales="free_y") +
    labs(x="Relative position from center", 
         y="Average Signal")
```

Meta-profile showing the effect of the exosome knockdown on eRNAs detected by CAGE:
```{r}
MM2 <- tidyMetaProfile(sites = by_clusterType, 
                      forward=CAGE_plus, reverse=CAGE_minus,
                      upstream=300, downstream=300)

head(MM2)

MM2 %>%
    gather(key="direction", value="score", sense, anti, factor_key=TRUE) %>%
    ggplot(aes(x=pos0, y=score, color=direction)) + 
    geom_line(alpha=0.75) + 
    facet_grid(sites~signal, scales="free_y") +
    scale_color_brewer("Direction", palette="Set1") +
    labs(x="Relative position from center", 
         y="Average Signal")
```

# Advanced usage

## Parallel Execution

In case multiple genomic signals are provided (Multiple `BigWigFile`-objects in a `BigWigFileList` or multiple `RleList`-objects in an `RleListList`), signals can be analyzed in parallel using the BiocParallel package. TidyMetaProfile uses the default registered backend:

```{r eval=FALSE}
library(BiocParallel)

# Set the backend to run calculations in series
register(SerialParam())

# Set the backend to run parallelize calculations using i.e. 3 cores:
register(MulticoreParam(workers=3))
```

## Alternative meta-summary functions

Instead of calculating the mean across sites, alternative summary functions can be provide. For example, instead of the average meta-profile, we can plot the median meta-profile:

```{r}
# Recalculate the first example using medians
SS1_median <- tidyMetaProfile(sites = promoters_only, 
                      forward=ChIP_Seq$DNase, reverse=NULL,
                      upstream=300, downstream=300,
                      sumFun = matrixStats::colMedians)

# Merge the two profiles and plot
list(mean=SS1, median=SS1_median) %>% 
    bind_rows(.id="summary") %>%
    ggplot(aes(x=pos0, y=sense, color=summary)) +
    geom_line(alpha=0.75) +
    geom_vline(xintercept=0, alpha=0.75, linetype="dotted") +
    scale_color_discrete("Summary-function") +
    labs(x="Basepair position relative to center",
         y="Average DNase signal")
```

## Outlier removal/trimming

In many cases, a few sites may have very extreme values which can disproportionally skew the calculated average profiles. tidyMetaProfile can automatically trim the sites with lowest and/or highest signals based on quantiles:

```{r}
# Recalculate the first example with different quantile trimmings:
SS1_95percent <- tidyMetaProfile(sites = promoters_only, 
                      forward=ChIP_Seq$DNase, reverse=NULL,
                      upstream=300, downstream=300,
                      trimUpper=0.95)

SS1_90percent <- tidyMetaProfile(sites = promoters_only, 
                      forward=ChIP_Seq$DNase, reverse=NULL,
                      upstream=300, downstream=300,
                      trimUpper=0.90)

# Merge the three profiles and plot
list(`100%`=SS1, 
     `95%`=SS1_95percent, 
     `90%`=SS1_90percent) %>% 
    bind_rows(.id="summary") %>%
    ggplot(aes(x=pos0, y=sense, color=summary)) +
    geom_line(alpha=0.75) +
    geom_vline(xintercept=0, alpha=0.75, linetype="dotted") +
    scale_color_discrete("Trimming-level") +
    labs(x="Basepair position relative to center",
         y="Average DNase signal")
```

## Access to low-level functions

The low-level functions used by tidyMetaProfile are all exposed for advanced user:

- `agnosticImport`: Import of chunks from BigWigFile or RleList
- `wideMetaProfile`: Output results as a site-by-position matrix.
- `quantileTrim`: Trim outliers from a single matrix or a pair of matrices.

The man pages for those functions contains details on how to use them. 

# Session Info

```{r}
sessionInfo()
```
